torch/csrc/api/src/data/samplers/distributed.cpp

```diff
@@ -0,0 +1,190 @@
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  auto iter = all_indices_.begin();
+  std::vector<size_t> res(iter + sample_index_, iter + end);
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedRandomSampler::reset(optional<size_t> new_size) {
+  size_t size = new_size.value_or(size_);
+  if (size != size_) {
+    size_ = size;
+    populate_indices();
+  }
+  std::minstd_rand rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
+  sample_index_ = begin_index_;
+}
+
+void DistributedRandomSampler::populate_indices() {
+  size_t sample_count =
+      num_replicas_ == 1 ? size_ : local_sample_count_ * num_replicas_;
+  all_indices_.resize(sample_count);
+  std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+  if (num_replicas_ > 1 && sample_count > size_) {
+    for (size_t i = size_; i < sample_count; ++i) {
+      all_indices_[i] =
+          i % size_; // we may have added duplicate samples to make all
+                     // replicas to have the same number of samples.
+    }
+  }
+
+  begin_index_ = rank_ * local_sample_count_;
+  end_index_ = begin_index_ + local_sample_count_;
+  sample_index_ = begin_index_;
+}
+
+void DistributedRandomSampler::save(serialize::OutputArchive& archive) const {
+  archive.write(
+      "sample_index_",
+      torch::tensor(static_cast<int64_t>(sample_index_), torch::kInt64),
+      /*is_buffer=*/true);
+}
+
+void DistributedRandomSampler::load(serialize::InputArchive& archive) {
+  auto tensor = torch::empty(1, torch::kInt64);
+  archive.read(
+      "sample_index_",
+      tensor,
+      /*is_buffer=*/true);
+  sample_index_ = tensor.item<int64_t>();
+}
+
+size_t DistributedRandomSampler::index() const noexcept {
+  return sample_index_;
+}
+
+DistributedSequentialSampler::DistributedSequentialSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : DistributedSampler(size, num_replicas, rank, allow_duplicates),
+      begin_index_(0),
+      end_index_(0),
+      sample_index_(0) {
+  populate_indices();
+}
+
+optional<std::vector<size_t>> DistributedSequentialSampler::next(
+    size_t batch_size) {
+  if (sample_index_ == end_index_) {
+    return nullopt;
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  std::vector<size_t> res(end - sample_index_);
+  std::iota(std::begin(res), std::end(res), sample_index_);
+  if (end >= size_) {
+    for (size_t& index : res) {
+      index = index % size_;
+    }
+  }
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedSequentialSampler::reset(optional<size_t> new_size) {
+  size_t size = new_size.value_or(size_);
+  if (size != size_) {
+    size_ = size;
+    populate_indices();
+  } else {
+    sample_index_ = begin_index_;
+  }
+}
+
+void DistributedSequentialSampler::populate_indices() {
+  begin_index_ = rank_ * local_sample_count_;
```

Comment:
I honestly don't get it why do we even keep this thing around. It's a single multiplication that you save at every `reset()`!

Reply:
This class could have easily be written by keeping only one or two of those variables.

Reply:
Yes, for this class we can do  the calculation in one place, but I don't want this calculation to repeat in all samplers again and again and hence it is calculated at DistributedSampler, now either remember this or use a method call. As i have mentioned, a method call require knowing whether to duplicate or  not.